File: File_IO.cpp
#include <iostream>
#include <fstream>
#include <chrono>
/*
This C++ program performs an I/O-bound benchmark by writing and reading a large number of integers to and from a file.
*/
int main() {
    const size_t N = 10'000'000; // number of integers
    const std::string filename = "io_test_c_plus_plus.txt";

    // Start timing
    auto start = std::chrono::high_resolution_clock::now();

    // Write numbers to file
    std::ofstream outFile(filename);
    for (size_t i = 0; i < N; ++i) {
        outFile << i << "\n";
    }
    outFile.close();

    // Read numbers and sum them
    std::ifstream inFile(filename);
    long long sum = 0;
    size_t value;
    while (inFile >> value) {
        sum += value;
    }
    inFile.close();

    // End timing
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    // Output results
    std::cout << "C++: I/O-bound benchmark with " << N << " numbers took "
              << elapsed.count() << " seconds.\n";
    std::cout << "Sum = " << sum << std::endl;

    return 0;
}

File: File_IO.py
import time

def main():
    N = 10_000_000
    filename = "io_test_python.txt"

    start = time.time()

    # Write numbers to file
    with open(filename, "w") as f:
        for i in range(N):
            f.write(f"{i}\n")

    # Read numbers and sum them
    total = 0
    with open(filename, "r") as f:
        for line in f:
            total += int(line.strip())

    end = time.time()

    print(f"Python: I/O-bound benchmark with {N} numbers took {end - start:.4f} seconds")
    print(f"Sum = {total}")

if __name__ == "__main__":
    main()
